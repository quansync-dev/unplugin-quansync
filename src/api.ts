export { transformQuansync } from './core/index'
