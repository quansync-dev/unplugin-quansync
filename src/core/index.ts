import {
  babelParse,
  getLang,
  isCallOf,
  isFunctionType,
  walkAST,
  walkImportDeclaration,
  type ImportBinding,
} from 'ast-kit'
import {
  generateTransform,
  MagicString,
  type CodeTransform,
} from 'magic-string-ast'
import type * as t from '@babel/types'

export function transformQuansync(
  code: string,
  id: string,
): CodeTransform | undefined {
  const lang = getLang(id)
  const program = babelParse(code, lang)
  const imports: Record<string, ImportBinding> = Object.create(null)

  for (const node of program.body) {
    if (node.type === 'ImportDeclaration') {
      walkImportDeclaration(imports, node)
    }
  }

  const macroName = Object.values(imports).find(
    (i) => i.source === 'quansync' && i.imported === 'quansyncMacro',
  )?.local
  if (!macroName) return

  const s = new MagicString(code)
  const functionScopes: boolean[] = []

  let inject = false
  function injectHelper() {
    if (inject) return
    inject = true
    s.prepend(`import { toGenerator as _QUANSYNC_TO_GEN } from 'quansync'\n`)
  }

  walkAST<t.Node>(program, {
    enter(node, parent) {
      if (node.type === 'AwaitExpression' && functionScopes.at(-1)) {
        injectHelper()
        s.overwrite(
          node.start!,
          node.argument.start!,
          'yield * _QUANSYNC_TO_GEN(',
        )
        s.appendLeft(node.end!, ')')
        return
      }

      if (!isFunctionType(node)) return

      const inMacroFunction = isCallOf(parent, macroName)
      functionScopes.push(inMacroFunction)

      if (!inMacroFunction || !node.async) return

      const name = 'id' in node && node.id ? node.id.name : ''
      const firstParam = node.params[0]
      const isArrowFunction = node.type === 'ArrowFunctionExpression'

      if (isArrowFunction) {
        if (firstParam) {
          s.overwrite(node.start!, firstParam.start!, `(`)
        } else {
          s.overwrite(node.start!, node.body.start!, `() => `)
        }

        if (node.body.type === 'BlockStatement') {
          s.appendLeft(node.body.start! + 1, `\nreturn function* () {`)
          s.appendLeft(node.body.end! - 1, `}.call(this)\n`)
        } else {
          throw new SyntaxError('Inlined arrow function is not supported')
        }
      } else if (firstParam) {
        s.overwrite(node.start!, firstParam.start!, `function* ${name}(`)
      } else {
        s.overwrite(node.start!, node.body.start!, `function* ${name}()`)
      }
    },
    leave(node) {
      if (isFunctionType(node)) {
        functionScopes.pop()
      }
    },
  })

  return generateTransform(s, id)
}
