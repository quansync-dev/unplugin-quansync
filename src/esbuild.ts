/**
 * This entry file is for esbuild plugin.
 *
 * @module
 */

import { Quansync } from './index'

/**
 * Esbuild plugin
 *
 * @example
 * ```ts
 * import { build } from 'esbuild'
 * import Quansync from 'unplugin-quansync/esbuild'
 * 
 * build({ plugins: [Quansync()] })
```
 */
const esbuild = Quansync.esbuild as typeof Quansync.esbuild
export default esbuild
export { esbuild as 'module.exports' }
