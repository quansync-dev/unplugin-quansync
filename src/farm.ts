/**
 * This entry file is for Farm plugin.
 *
 * @module
 */

import { Quansync } from './index'

/**
 * Farm plugin
 *
 * @example
 * ```ts
 * // farm.config.js
 * import Quansync from 'unplugin-quansync/farm'
 *
 * export default {
 *   plugins: [Quansync()],
 * }
 * ```
 */
const farm = Quansync.farm as typeof Quansync.farm
export default farm
export { farm as 'module.exports' }
