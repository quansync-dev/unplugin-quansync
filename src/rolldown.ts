/**
 * This entry file is for Rolldown plugin.
 *
 * @module
 */

import { Quansync } from './index'

/**
 * Rolldown plugin
 *
 * @example
 * ```ts
 * // rolldown.config.js
 * import Quansync from 'unplugin-quansync/rolldown'
 *
 * export default {
 *   plugins: [Quansync()],
 * }
 * ```
 */
const rolldown = Quansync.rolldown as typeof Quansync.rolldown
export default rolldown
export { rolldown as 'module.exports' }
