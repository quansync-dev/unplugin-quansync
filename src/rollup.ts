/**
 * This entry file is for Rollup plugin.
 *
 * @module
 */

import { Quansync } from './index'

/**
 * Rollup plugin
 *
 * @example
 * ```ts
 * // rollup.config.js
 * import Quansync from 'unplugin-quansync/rollup'
 *
 * export default {
 *   plugins: [Quansync()],
 * }
 * ```
 */
const rollup = Quansync.rollup as typeof Quansync.rollup
export default rollup
export { rollup as 'module.exports' }
