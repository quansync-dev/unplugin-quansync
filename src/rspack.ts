/**
 * This entry file is for Rspack plugin.
 *
 * @module
 */

import { Quansync } from './index'

/**
 * Rspack plugin
 *
 * @example
 * ```js
 * // rspack.config.js
 * import Quansync from 'unplugin-quansync/rspack'
 *
 * default export {
 *  plugins: [Quansync()],
 * }
 * ```
 */
const rspack = Quansync.rspack as typeof Quansync.rspack
export default rspack
export { rspack as 'module.exports' }
