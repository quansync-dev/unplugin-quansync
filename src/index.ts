import { createUnplugin, type UnpluginInstance } from 'unplugin'
import { transformQuansync } from './core'
import { resolveOptions, type Options } from './core/options'

export const Quansync: UnpluginInstance<Options | undefined, false> =
  createUnplugin((rawOptions = {}) => {
    const options = resolveOptions(rawOptions)

    const name = 'unplugin-quansync'
    return {
      name,
      enforce: options.enforce,

      transform: {
        filter: {
          id: {
            include: options.include,
            exclude: options.exclude,
          },
          code: 'quansync',
        },
        handler(code, id) {
          return transformQuansync(code, id)
        },
      },
    }
  })
