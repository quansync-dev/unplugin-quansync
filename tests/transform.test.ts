/// <reference types="vite/client" />
import { writeFile } from 'node:fs/promises'
import path from 'node:path'
import { pathToFileURL } from 'node:url'
import { testFixtures } from '@sxzz/test-utils'
import { describe } from 'vitest'
import { transformQuansync } from '../src/core'

describe('transform', async () => {
  const dirname = import.meta.dirname
  await testFixtures(
    import.meta.glob<string>('./fixtures/*.js', {
      eager: true,
      query: '?raw',
      import: 'default',
    }),
    async (args, id, code) => {
      const result = transformQuansync(code, id)?.code
      if (!result) return result

      const filePath = path.resolve(dirname, `temp/${path.basename(id)}`)
      await writeFile(filePath, result)
      const mod = await import(`${pathToFileURL(filePath).href}?${Date.now()}`)
      await mod.default?.()

      return result
    },
    { promise: true },
  )
})
