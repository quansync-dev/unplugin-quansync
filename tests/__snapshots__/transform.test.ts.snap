// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > ./fixtures/alias.js 1`] = `
"// @ts-check
import { quansync as x } from 'quansync/macro'
import { expect } from 'vitest'

export const echo = x({
  sync: /** @param {string} v */ (v) => v,
  async: (v) => Promise.resolve(v),
})

const echoNewLine = x(
  /** @param {string|Promise<string>} v */ function* (v) {
    const contents = yield echo(yield v)
    return contents + '\\n'
  },
)

export default async () => {
  const iter = echoNewLine('hello')
  expect(iter).a('Generator')
  await expect(iter).resolves.toBe('hello\\n')
  expect(echoNewLine.sync('world')).toBe('world\\n')

  await expect(echoNewLine(Promise.resolve('hello'))).resolves.toBe('hello\\n')
  expect(() => echoNewLine.sync(Promise.resolve('world'))).throw()
}
"
`;

exports[`transform > ./fixtures/basic.js 1`] = `
"// @ts-check
import { quansync } from 'quansync/macro'
import { expect } from 'vitest'

export const getNumber = quansync({
  sync: /** @param {number} id */ (id) => id,
  async: (id) => Promise.resolve(id),
})

const inc1 = quansync(function* (){
  const value = yield getNumber(1)
  return value + 1
})

const inc2 = quansync(function* (id) {
  const value = yield getNumber(1)
  return value + 1
})

const inc3 = quansync(function* named(){
  const value = yield getNumber(1)
  return value + 1
})

const inc4 = quansync(function* named(id) {
  const value = yield getNumber(1)
  return value + 1
})

const inc5 = quansync(() => {
return function* () {
  const value = yield getNumber(1)
  return value + 1
}.call(this)
})

const inc6 = quansync((id) => {
return function* () {
  const value = yield getNumber(1)
  return value + 1
}.call(this)
})

export const fn7 = quansync(() => {
return function* () {
  yield 1
  yield 2
  const fn = async () => {
    await 3
  }
  async function x() {
    await 4
  }
  class Foo {
    async x() {
      await 10
    }
    async #x() {
      await 10
    }
  }
}.call(this)
})

quansync(() => {
return function* () {
  ;!(yield 1)
  ;'' + (yield 1)
  ;false && (yield 1)
  fn7()
  ;(yield 1) < (yield 10)
  const x = (yield 1) < (yield 10)
}.call(this)
})

export default async () => {
  await expect(getNumber(1)).resolves.toBe(1)
  expect(getNumber.sync(1)).toBe(1)
  for (const fn of [inc1, inc2, inc3, inc4, inc5, inc6]) {
    await expect(fn()).resolves.toBe(2)
    expect(fn.sync(1)).toBe(2)
  }
}
"
`;

exports[`transform > ./fixtures/inlined-arrow.js 1`] = `
"// @ts-check
import { quansync } from 'quansync/macro'
import { expect } from 'vitest'

export const echo = quansync({
  sync: /** @param {string} v */ (v) => v,
  async: (v) => Promise.resolve(v),
})

const echoNewLine = quansync(
  /** @param {string|Promise<string>} v */ (v) =>
    {
return function* () {
return ((yield echo(yield v))) + '\\n'
}.call(this)
},
)

export default async () => {
  const iter = echoNewLine('hello')
  expect(iter).a('Generator')
  await expect(iter).resolves.toBe('hello\\n')
  expect(echoNewLine.sync('world')).toBe('world\\n')

  await expect(echoNewLine(Promise.resolve('hello'))).resolves.toBe('hello\\n')
  expect(() => echoNewLine.sync(Promise.resolve('world'))).throw()
}
"
`;
